### ISSUE TITLE
[no-unused-vars] type information of 'this' variable lost within forEach

### ISSUE THREAD

**Issue by mwoodpatrick** [typescript-eslint/typescript-eslint#4355]
- [ y] I have tried restarting my IDE and the issue persists.
- [ y] I have updated to the latest version of the packages.
- [ y] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

In the repro below when I point at the 'this' at the start of the 'process' method vscode can find the type definition but inside the 'forEach' it can't find the type definition!  eslint reports correctly that 'target' is defined but never used. but the reference to 'this.target' should be reported. Does anyone have any suggestions on the best way to debug or what the issue is, or what I'm doing wrong!

**Repro**

```ts
class myclass {
  name: string;
  targets: string[];

  process() {
    this.targets.forEach(function (target) {
      console.log(`${this.name} and ${this.target}`);
    });
  }
}

const x = new myclass();

x.process();
```

**Expected Result**

Should report that 'target' does not exist in 'myclass' and that 'target' is unused

**Actual Result**

Does report that 'target' is unused but does not report that  'target' does not exist in 'myclass

**Additional Info**

See attached output from eslint & tsconfig files

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | 5.3.1 |
| `@typescript-eslint/parser`        | `5.3.1` |
| `TypeScript`                       | 4.4.4 |
| `ESLint`                           | 8.2.0 |
| `node`                             |v16.13.1 |

**Comment by JoshuaKGoldberg** [comment]
Hi @mwoodpatrick! When you use a traditional function expression (`function`) instead of an arrow function (`() =>`), the `this` is not preserved. Compare [this TypeScript playground example with your code](https://www.typescriptlang.org/play?strictPropertyInitialization=false#code/MYGwhgzhAEC2CepIwN4ChrQHZlgUwC5oIAXAJwEssBzAbg2hLDOrxIiNMpoG0BdegwAOZAPbA8UABQBKaOkyYSACwoQAdExZsNAM1FkAomGDKpugK5ZgJCqKzQpW1iTkLFmYPYiiQedSCi1FIABgAkKCpq6jj4AL7QYFgAJtARURrObHEhMvQecXkMcWglaF5YpNAAHtAAvNh4AO5wiODSRWjV6iLikhCy9EA) to [this one with a fix](https://www.typescriptlang.org/play?strictPropertyInitialization=false#code/MYGwhgzhAEC2CepIwN4ChrQHZlgUwC5oIAXAJwEssBzAbg2hLDOrxIiNMpoG0BdegwAOZAPbA8UABQBKaOkyYSACwoQAdExZsNAM1FkAomGDKpUraxJyAvAD55DRdGCisEUSDzqQo6lIADABIUFTV1HHwAX2gwLAATaBDLNiiAmXpnKIyGKLQ8tFd3EmgAD2gbbDwAdzhEcGkctFL1EXFJCFl6IA).

In general, I'd suggest asking for help on the [TypeScript Community Discord](https://discord.com/invite/typescript) if you're not sure whether you're doing something wrong. Most of the time it is an issue in user code, not ESLint, TypeScript, or typescript-eslint. This issue tracker isn't intended to be a support forum (and wouldn't be a very discoverable or good one).

By the way - I took the liberty of editing your post to properly use triple backticks to format code. A little more readable 🙂. Cheers!
